File: algorithms/2_medium/M/00318_maximum_product_of_word_lengths/bit_manipulation_bruteforce.py
class Solution:
    def maxProduct(self, words: List[str]) -> int:
        n = len(words)
        masks = [0] * n

        for i in range(n):
            mask = 0

            for ch in words[i]:
                mask = mask | (1 << (ord(ch) - ord('a')))

            masks[i] = mask

        max_product_len = 0

        for i in range(n):
            for j in range(i + 1, n):
                if (masks[i] & masks[j]) == 0:
                    max_product_len = max(max_product_len, len(words[i]) * len(words[j]))

        return max_product_len

File: algorithms/2_medium/M/00318_maximum_product_of_word_lengths/bit_manipulation_bruteforce.rs
impl Solution {
    pub fn max_product(words: Vec<String>) -> i32 {
        let len = words.len();
        let mut masks = vec![0_u32; len];

        for i in 0..len {
            let mut mask = 0_u32;

            for &byte in words[i].as_bytes().iter() {
                mask = mask | (1 << (byte - b'a'));
            }

            masks[i] = mask;
        }

        let mut max_product_len = 0_usize;

        for i in 0..len {
            for j in (i + 1)..len {
                if (masks[i] & masks[j]) == 0 {
                    max_product_len = max_product_len.max(words[i].len() * words[j].len());
                }
            }
        }

        max_product_len as i32
    }
}
